core/src/main/java/kafka/server/share/ShareFetchUtils.java

```diff
@@ -44,61 +45,61 @@
-                        // replica manager fetch. Depends on the share partition manager implementation,
-                        // if we want parallel requests for the same share partition or not.
-                        partitionData.setPartitionIndex(topicIdPartition.partition())
-                                .setRecords(fetchPartitionData.records)
-                                .setErrorCode(fetchPartitionData.error.code())
-                                .setAcquiredRecords(acquiredRecords)
-                                .setAcknowledgeErrorCode(Errors.NONE.code());
-                        return partitionData;
-                    }));
-        });
-        return CompletableFuture.allOf(futures.values().toArray(new CompletableFuture[0])).thenApply(v -> {
-            Map<TopicIdPartition, ShareFetchResponseData.PartitionData> processedResult = new HashMap<>();
-            futures.forEach((topicIdPartition, future) -> processedResult.put(topicIdPartition, future.join()));
-            return processedResult;
+                // In case we get OFFSET_OUT_OF_RANGE error, that's because the Log Start Offset is later than the fetch offset.
+                // So, we would update the start and end offset of the share partition and still return an empty
+                // response and let the client retry the fetch. This way we do not lose out on the data that
+                // would be returned for other share partitions in the fetch request.
+                if (fetchPartitionData.error.code() == Errors.OFFSET_OUT_OF_RANGE.code()) {
+                    sharePartition.updateCacheAndOffsets(offsetForEarliestTimestamp(topicIdPartition, replicaManager));
+                    // We set the error code to NONE, as we have updated the start offset of the share partition
+                    // and the client can retry the fetch.
+                    partitionData.setErrorCode(Errors.NONE.code());
+                }
+            } else {
+                List<AcquiredRecords> acquiredRecords = sharePartition.acquire(shareFetchData.memberId(), fetchPartitionData);
+                log.trace("Acquired records for topicIdPartition: {} with share fetch data: {}, records: {}",
+                    topicIdPartition, shareFetchData, acquiredRecords);
+                // Maybe, in the future, check if no records are acquired, and we want to retry
+                // replica manager fetch. Depends on the share partition manager implementation,
+                // if we want parallel requests for the same share partition or not.
+                if (acquiredRecords.isEmpty()) {
+                    partitionData
+                        .setRecords(null)
+                        .setAcquiredRecords(Collections.emptyList());
+                } else {
+                    partitionData
+                        .setRecords(fetchPartitionData.records)
+                        .setErrorCode(fetchPartitionData.error.code())
```

Comment:
No need to set error code here since we don't set it in the `if` path?

Reply:
Yeah, my bad. Done.